domain/src/main/java/org/oppia/android/domain/spotlight/SpotlightStateController.kt

```diff
@@ -0,0 +1,182 @@
+   * @param feature the spotlight feature who's view state is to be recorded
+   * @throws SpotlightFeatureNotFoundException when feature is not set correctly
+   */
+  fun markSpotlightViewed(
+    profileId: ProfileId,
+    feature: Spotlight.FeatureCase,
+  ): DataProvider<Any?> {
+    val deferred = recordSpotlightStateAsync(
+      profileId,
+      feature,
+      SpotlightViewState.SPOTLIGHT_SEEN
+    )
+    return dataProviders.createInMemoryDataProviderAsync(
+      RECORD_SPOTLIGHT_CHECKPOINT_DATA_PROVIDER_ID
+    ) {
+      return@createInMemoryDataProviderAsync AsyncResult.Success(deferred.await())
+    }
+  }
+
+  /**
+   * Retrieves the current [SpotlightViewState] of a spotlit feature for a given profile.
+   *
+   * @param profileId the ID of the profile that will be viewing the spotlight
+   * @param feature the spotlight feature to be spotlit
+   *
+   * @return DataProvider containing the current [SpotlightViewState] corresponding to the specified [feature]
+   */
+  fun retrieveSpotlightViewState(
+    profileId: ProfileId,
+    feature: Spotlight.FeatureCase,
+  ): DataProvider<SpotlightViewState> {
+    return retrieveCacheStore(profileId)
+      .transformAsync(
+        RETRIEVE_SPOTLIGHT_CHECKPOINT_DATA_PROVIDER_ID
+      ) {
+
+        val viewState = when (feature) {
+          Spotlight.FeatureCase.ONBOARDING_NEXT_BUTTON -> {
+            it.onboardingNextButton
+          }
+          Spotlight.FeatureCase.TOPIC_LESSON_TAB -> {
+            it.topicLessonTab
+          }
+          Spotlight.FeatureCase.TOPIC_REVISION_TAB -> {
+            it.topicRevisionTab
+          }
+          Spotlight.FeatureCase.FIRST_CHAPTER -> {
+            it.firstChapter
+          }
+          Spotlight.FeatureCase.PROMOTED_STORIES -> {
+            it.promotedStories
+          }
+          Spotlight.FeatureCase.LESSONS_BACK_BUTTON -> {
+            it.lessonsBackButton
+          }
+          Spotlight.FeatureCase.VOICEOVER_PLAY_ICON -> {
+            it.voiceoverPlayIcon
+          }
+          Spotlight.FeatureCase.VOICEOVER_LANGUAGE_ICON -> {
+            it.voiceoverLanguageIcon
+          }
+          Spotlight.FeatureCase.FEATURE_NOT_SET -> {
+            return@transformAsync AsyncResult.Failure(
+              SpotlightFeatureNotFoundException("Spotlight feature was not found")
+            )
+          }
+        }
+        AsyncResult.Success(viewState)
+      }
+  }
+
+  private fun recordSpotlightStateAsync(
+    profileId: ProfileId,
+    feature: Spotlight.FeatureCase,
+    viewState: SpotlightViewState
+  ): Deferred<Any> {
+    return retrieveCacheStore(profileId).storeDataWithCustomChannelAsync(
+      updateInMemoryCache = true
+    ) {
+      val spotlightStateDatabaseBuilder = it.toBuilder()
+
+      val spotlight = when (feature) {
+        Spotlight.FeatureCase.ONBOARDING_NEXT_BUTTON -> {
+          spotlightStateDatabaseBuilder.setOnboardingNextButton(viewState)
+        }
```

Comment:
Per my comment above, if a custom channel isn't needed then you can probably simplify this by doing something like:

```kotlin
spotlightState.toBuilder().run { builder ->
  when (feature) {
    ONBOARDING_NEXT_BUTTON -> builder.onboardingNextButton = viewState
    ...
  }
}.build()
```

The above is a bit more compacted and still requires an exhaustive ``when`` statement.

Reply:
done